### PR TITLE
 Remove all usages of numpy.compat ref #13880

### PR DESCRIPTION
**Pull request by harishsdev**
#### Reference issue

https://github.com/scipy/scipy/issues/13880

#### What does this implement/fix?
Remove all usages of numpy.compat(for asbytes and asstr functionality)

### CODE REVIEW

File: scipy/_lib/_util.py [hunk 1]
```diff
@@ -14,7 +14,14 @@
...
         pass
-
+def asstr(s):
+    if isinstance(s, bytes):
+        return s.decode('latin1')
+    return str(s)
+def asbytes(s):
+    if isinstance(s, bytes):
+        return s
+    return str(s).encode('latin1')
```
Comment (rgommers) [inline scipy/_lib/_util.py:24]: These are py2-py3 compatibility functions. Given that we only use Python 3, we should not need them. When removing them, the idea is to actually figure out if something is a string or bytes. 

I'll give you one example for `idl.py`.

File: scipy/io/idl.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
 
 import struct
 import numpy as np
-from numpy.compat import asstr
+from scipy._lib._util import asstr
```
Comment (rgommers) [inline scipy/io/idl.py:34]: You can remove this import. Then look at where it is used. In this case in one place:
```
    if length > 0:
        chars = _read_bytes(f, length)
        _align_32(f)
        chars = asstr(chars)
    else:
        chars = ''
```

Now, you look at the file mode with which `f` was opened. In this case that was `f = open(file_name, 'rb')`. So `f.read()` is guaranteed to give `bytes`. Hence you can change the code to:
```
    if length > 0:
        chars = _read_bytes(f, length).decode('latin1')
        _align_32(f)
    else:
        chars = ''
```

Then you run the tests to confirm - without `.decode('latin1')` they all fail, with the decode added they all pass.

**Review by rgommers — CHANGES_REQUESTED** [review]
Hi @harishsdev, thanks for the PR. This isn't quite right - I explained in more detail what to do.

Also, please write your code in PEP8 compatible fashion (e.g., spaces after commas).

**Comment by harishsdev** [comment]
i tried to run tests using following command but i am facing issue,

if its exists,please guide me how to proceed


(pyart_env) PS D:\one\scipy> python .\runtests.py
Building, see build.log...
scipy/linalg/_generate_pyx.py: all files up-to-date
scipy/special/_generate_pyx.py: all files up-to-date
Running scipy\linalg\_generate_pyx.py
Running scipy\special\_generate_pyx.py
Running scipy\stats\_generate_pyx.py
scipy\cluster\_hierarchy.pyx has not changed
scipy\io\matlab\mio_utils.pyx has not changed
scipy\io\matlab\streams.pyx has not changed
scipy\linalg\cython_blas.pyx has not changed
scipy\linalg\cython_lapack.pyx has not changed
scipy\interpolate\_bspl.pyx has not changed
scipy\linalg\_matfuncs_sqrtm_triu.pyx has not changed
scipy\linalg\_solve_toeplitz.pyx has not changed
scipy\ndimage\src\_cytest.pyx has not changed
scipy\ndimage\src\_ni_label.pyx has not changed
scipy\optimize\_bglu_dense.pyx has not changed
scipy\optimize\_group_columns.pyx has not changed
scipy\linalg\_decomp_update.pyx.in has not changed
scipy\optimize\_highs\cython\src\_highs_constants.pyx has not changed
scipy\optimize\_highs\cython\src\_highs_wrapper.pyx has not changed
scipy\optimize\_lsq\givens_elimination.pyx has not changed
scipy\optimize\_trlib\_trlib.pyx has not changed
scipy\signal\_max_len_seq_inner.pyx has not changed
scipy\signal\_peak_finding_utils.pyx has not changed
scipy\signal\_sosfilt.pyx has not changed
scipy\signal\_spectral.pyx has not changed
scipy\signal\_upfirdn_apply.pyx has not changed
scipy\sparse\_csparsetools.pyx.in has not changed
scipy\sparse\csgraph\_flow.pyx has not changed
scipy\sparse\csgraph\_matching.pyx has not changed
scipy\sparse\csgraph\_min_spanning_tree.pyx has not changed
scipy\sparse\csgraph\_reordering.pyx has not changed
scipy\sparse\csgraph\_shortest_path.pyx has not changed
scipy\sparse\csgraph\_tools.pyx has not changed
scipy\sparse\csgraph\_traversal.pyx has not changed
scipy\spatial\ckdtree.pyx has not changed
scipy\spatial\qhull.pyx has not changed
scipy\spatial\_hausdorff.pyx has not changed
scipy\spatial\_voronoi.pyx has not changed
scipy\spatial\transform\rotation.pyx has not changed
scipy\special\cython_special.pyx has not changed
scipy\special\_comb.pyx has not changed
scipy\special\_ellip_harm_2.pyx has not changed
scipy\special\_test_round.pyx has not changed
scipy\special\_ufuncs.pyx has not changed
scipy\special\_ufuncs_cxx.pyx has not changed
scipy\stats\biasedurn.pyx has not changed
scipy\stats\_qmc_cy.pyx has not changed
scipy\stats\_sobol.pyx has not changed
scipy\stats\_stats.pyx has not changed
scipy\_lib\messagestream.pyx has not changed
scipy\_lib\_ccallback_c.pyx has not changed
scipy\_lib\_test_deprecation_call.pyx has not changed
scipy\_lib\_test_deprecation_def.pyx has not changed
scipy\io\matlab\mio5_utils.pyx has not changed
scipy\cluster\_optimal_leaf_ordering.pyx has not changed
scipy\cluster\_vq.pyx has not changed
scipy\fftpack\convolve.pyx has not changed
scipy\interpolate\interpnd.pyx has not changed
scipy\optimize\cython_optimize\_zeros.pyx.in has not changed
scipy\interpolate\_ppoly.pyx has not changed
Cythonizing sources
lapack_opt_info:
lapack_mkl_info:
No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
customize MSVCCompiler
  libraries mkl_rt not found in ['G:/install/envs/pyart_env\\Library\\lib']
  NOT AVAILABLE

openblas_lapack_info:
  libraries openblas not found in ['G:/install/envs/pyart_env\\Library\\lib']
get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'
customize GnuFCompiler
Could not locate executable g77
Could not locate executable f77
customize IntelVisualFCompiler
Could not locate executable ifort
Could not locate executable ifl
customize AbsoftFCompiler
Could not locate executable f90
customize CompaqVisualFCompiler
Could not locate executable DF
customize IntelItaniumVisualFCompiler
Could not locate executable efl
customize Gnu95FCompiler
Could not locate executable gfortran
Could not locate executable f95
customize G95FCompiler
Could not locate executable g95
customize IntelEM64VisualFCompiler
customize IntelEM64TFCompiler
Could not locate executable efort
Could not locate executable efc
customize PGroupFlangCompiler
Could not locate executable flang
don't know how to compile Fortran code on platform 'nt'
  NOT AVAILABLE

openblas_clapack_info:
  libraries openblas,lapack not found in ['G:/install/envs/pyart_env\\Library\\lib']
  NOT AVAILABLE

flame_info:
  libraries flame not found in ['G:/install/envs/pyart_env\\Library\\lib']
  NOT AVAILABLE

atlas_3_10_threads_info:
Setting PTATLAS=ATLAS
  libraries lapack_atlas not found in G:/install/envs/pyart_env\Library\lib
  libraries tatlas,tatlas not found in G:/install/envs/pyart_env\Library\lib
<class 'numpy.distutils.system_info.atlas_3_10_threads_info'>
  NOT AVAILABLE

atlas_3_10_info:
  libraries lapack_atlas not found in G:/install/envs/pyart_env\Library\lib
  libraries satlas,satlas not found in G:/install/envs/pyart_env\Library\lib
<class 'numpy.distutils.system_info.atlas_3_10_info'>
  NOT AVAILABLE

atlas_threads_info:
Setting PTATLAS=ATLAS
  libraries lapack_atlas not found in G:/install/envs/pyart_env\Library\lib
  libraries ptf77blas,ptcblas,atlas not found in G:/install/envs/pyart_env\Library\lib
<class 'numpy.distutils.system_info.atlas_threads_info'>
  NOT AVAILABLE

atlas_info:
  libraries lapack_atlas not found in G:/install/envs/pyart_env\Library\lib
  libraries f77blas,cblas,atlas not found in G:/install/envs/pyart_env\Library\lib
<class 'numpy.distutils.system_info.atlas_info'>
  NOT AVAILABLE

lapack_info:
  libraries lapack,blas not found in ['G:/install/envs/pyart_env\\Library\\lib']
  NOT AVAILABLE

lapack_src_info:
  NOT AVAILABLE

  NOT AVAILABLE

Running from SciPy source directory.
G:\install\envs\pyart_env\lib\site-packages\numpy\distutils\system_info.py:1849: UserWarning:
    Lapack (http://www.netlib.org/lapack/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [lapack]) or by setting
    the LAPACK environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
G:\install\envs\pyart_env\lib\site-packages\numpy\distutils\system_info.py:1849: UserWarning:
    Lapack (http://www.netlib.org/lapack/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [lapack_src]) or by setting
    the LAPACK_SRC environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
Traceback (most recent call last):
  File "setup.py", line 619, in <module>
    setup_package()
  File "setup.py", line 615, in setup_package
    setup(**metadata)
  File "G:\install\envs\pyart_env\lib\site-packages\numpy\distutils\core.py", line 135, in setup
    config = configuration()
  File "setup.py", line 518, in configuration
    raise NotFoundError(msg)
numpy.distutils.system_info.NotFoundError: No BLAS/LAPACK libraries found.
To build Scipy from sources, BLAS & LAPACK libraries need to be installed.
See site.cfg.example in the Scipy source directory and
https://docs.scipy.org/doc/scipy/reference/building/index.html for details.

Build failed! (0:00:01.512716 elapsed)

**Review by peterbell10 — CHANGES_REQUESTED** [review]
The log tells you what's wrong at the bottom:

>numpy.distutils.system_info.NotFoundError: No BLAS/LAPACK libraries found.
To build Scipy from sources, BLAS & LAPACK libraries need to be installed.

Perhaps this windows guide will help with installing libraries:
https://docs.scipy.org/doc/scipy/reference/building/windows.html#build-windows

File: scipy/io/matlab/mio5_utils.pyx [hunk 3]
```diff
@@ -918,7 +916,7 @@ cdef class VarReader5:
             char *n_ptr = names
             int j, dup_no
         for i in range(n_names):
-            name = asstr(PyBytes_FromString(n_ptr))
+            name = PyBytes_FromString(n_ptr).decode('latin1')
```
[the remark this answers is not in the document]
Comment (NamamiShanker) [reply scipy/io/matlab/mio5_utils.pyx]: This change is causing an error `'cpython/PyUnicode_FromString.pxd' not found`
How can I fix this?

**Comment by peterbell10** [comment]
Thanks for your contribution @harishsdev but the `numpy.compat` uses have now been replaced by gh-14009, so closing this PR.

**Pull request closed without merging**